### DISCUSSION TITLE
How to exclude non moving object, stationary objects?

### DISCUSSION THREAD

**Discussion by MathiasThinsz** [blakeblackshear/frigate#1128] [Ask A Question]
In front of my house I have a parking space. I'd like to be notified when a car stops there or drives away, but now Frigate is reporting that there is a car parked there. I don't want to mask out the parking lot since I really would like to know when a car arrives or departs. Is it possible to only detect moving items in the picture?

I've tried to search this forum for an answer, but haven't found any.

**Comment by blakeblackshear** [comment]
Generally people accomplish this by configuring zones and setting up alerts based on zones. You can also add zone requirements to snapshots and clips so frigate doesn't save events for a car in unwanted areas. Masks are designed to ignore motion in specific areas and/or to indicate areas that are guaranteed false positives.

Frigate is designed to track objects as they move and over masking can prevent it from knowing that an object in the current frame is the same as the previous frame.

You want frigate to detect objects everywhere and configure your snapshots, clips, and alerts to be based on the location of the object with zones.

**Reply by MathiasThinsz** [reply]
I see. But if Frigate is design to track object as they move, why is it reacting on things that are stationary? It would be a quite easy feature to be able to turn off the recognition for stationary objects, wouldn't it?
I want Frigate to detect when a car arrives and departs, ie when there is movement. But I don't need to have hundreds of clips saved showing me the parked car. 
Frigate is great but I think this is a huge disadvantage.

**Reply by blakeblackshear** [reply]
Motion detection works by determining if a sufficient number of pixels have changed between frames. Shadows or other lighting changes will be detected as motion. You can watch the feed to see where motion boxes occur and tweak the settings to make it less sensitive.

> It would be a quite easy feature to be able to turn off the recognition for stationary objects, wouldn't it?

Definitely not. It's on the todo list, but it is not easy to figure out if something is "stationary" across a stream of frames.

>  I don't need to have hundreds of clips saved showing me the parked car

This can be accomplished with zones and restricting snapshots and clips to objects that entered certain zones as I mentioned above.

**Reply by KillahB33** [reply]
I am spit balling here but would you be able to take the frames that were used to detect motion and isolate the object detection to the moved frames?

**Reply by blakeblackshear** [reply]
That's how it already works.

**Reply by KillahB33** [reply]
So I guess the issue he is trying to solve is if the light changes on the stationary object it will be detected as moving.
My bad! Great work btw! Coming from Shinobi and this is a lot cleaner!

**Comment by gadget-man** [comment]
I think my question is related to this one so posting it here: If I have a car parked on the driveway and a person walks into the zone, I get both a 'car' and a 'person' object detection. Is there a way to filter this in some way in HA to distinguish between a car arriving and the car being there when some other object arrives / moves?

**Reply by tve** [reply]
Sounds tricky... Maybe having a way to reduce the region size so it's closer to the motion bounding box could help. Or take the bounding boxes of the detected objects and exclude the ones that don't significantly overlap the motion boxes. But these would all be code changes...

Something that might work now is to filter the events that frigate emits via MQTT, I see that each event is for one object type and has a `current_zones` and an `entered_zones` field, so if you arrange zones so cars and people that should trigger need to go through two zones you could filter out the ones that only have one in `entered_zones`?

**Reply by gadget-man** [reply]
Thank you. Having looked more closely at the MQTT event messages, I think a combination of the `before` `current_zones` and `after` `entered_zones` may well be the answer - so I can ignore events when `car` is detected as the object but the `before` state already shows it as being present in the target zone.

**Reply by gadget-man** [reply]
That doesn't quite work as planned. A person walking up to an existing car provides exactly the same combination of `before` and `after` `current_zones` and `entered_zones`, so it seems there is no distinction between an object that's already present and a newly arrived object

**Reply by blakeblackshear** [reply]
Can you clarify what you mean by "filter this in some way in HA"? Are you trying to avoid notifications for the parked car?

**Reply by gadget-man** [reply]
Exactly that. I'm successfully detecting when a car arriving, but I then get multiple additional events of a car being detected when movement happens elsewhere in the same zone e.g. a person walking up to the car. I'm trying to avoid getting notifications for the parked car when the movement is related to a different event.

**Reply by blakeblackshear** [reply]
There are a few ways to get what you want. Create multiple zones that cover your driveway. For cars, you would only notify if entered_zones has more than 1 zone. For person, you would notify regardless of the number of entered_zones. You can also create a zone for the entrance of your driveway and only notify if that zone is in the list of entered_zones when the object is a car.

**Reply by gadget-man** [reply]
Ok that's helpful, thanks. Sounds like an obvious question, but is there any docs on the difference between 'entered_zones' and 'current_zones' and what causes a zone to be flagged in each?

**Reply by blakeblackshear** [reply]
`current_zones` are the zones the object is in right now. `entered_zones` are the zones the object as ever entered while being tracked.

**Reply by joaoubaldo** [reply]
I have the same issue with cars. Thought about a possible solution: trigger events only when the detected object didn't change its position (with some adjustable pixel margin of course) since last detection -- is this acceptable @blakeblackshear ?

**Reply by blakeblackshear** [reply]
Once motion stops, the event ends. When new motion occurs the object is detected as an entirely new object. There isn't an easy way to know that the object is the same across those events.

The best way is to do what I mentioned above.

**Comment by MarkGodwin** [comment]
I think the solution to this might be fairly simple with some changes to Frigate and the way it publishes the count of tracked objects to home assistant. It simply only needs to update the count of detected objects after it does an object detection pass (i.e. after being triggered by motion). If there is no motion detected, frigate shouldn't change the number of tracked objects.

If there is no motion, we can safely assume that the objects it detected before are still there. So if there is motion (say, a person) and the detection finds the same parked car as before, the object count would be the same as before, so would result in no car notification.

The _problem_ is frigate uses the same object count published number to identify motion as well as count the number of objects of a certain type. This means the object count has to go back to zero, so that motion is no longer reported. But really you could have no motion while objects are still present, but there is no way to represent this in the mqtt state. So this means the mqtt data needs two values - one for whether motion of that object type was detected recently, and another for how many objects were detected last time the detection was attempted.

This wouldn't be perfect, because you'd still get "car motion" notifications if a person walked past a parked car, for example, but at least the number of detected cars would remain unchanged, so the car count could easily be used for presence detection.

@blakeblackshear Would you consider accepting a PR along those lines? It's quite a big change to the MQTT stream, so would need an associated change to the HA addin. Backward compatibility of the MQTT messages would be achievable, though.

**Reply by blakeblackshear** [reply]
Have you looked at the code? It doesn't work like that at all. There is no concept of a motion count that is related to detected objects. Older versions used to track objects even after motion stopped. This became problematic when clips were introduced because the events would continue for a very long time and the cache would stay full until the object left. It also caused frigate to continuously run object detection. I made the decision to end events when motion stopped because it was better for the vast majority of users. 

The next release lays the groundwork for allowing objects to be tracked for long periods of time again by decoupling events and recordings/clips.

This will help, but there will still be edge cases where an object is lost and then redetected. Another solution I have thought about is tracking the distance in pixels that the object has moved. It would be a little complex because that distance means different things depending on where the object is in the frame, but I think it would be doable.

**Reply by gadget-man** [reply]
I think in the vast majority of cases the existing concept of zones should work fine. If movement is detected in 1 zone but not in another, then we know the object hasn't moved between zones so may not be actually 'moving' at all.   The only reason this doesn't work in my current scenario is because the current camera position means the car fills much of the frame when parked - including blocking out the 'entry' zone.   That's probably more a case of poor camera positioning for testing purposes than any problems with the underlying Frigate code, which is a hugely impressive piece of work @blakeblackshear!

**Reply by blakeblackshear** [reply]
@gadget-man don't forget that the bottom center of the bounding box is used to evaluate whether or not an object is in the zone, not the entire bounding box. You can probably get more granular than you think even with your car filling the frame.

Another way to measure movement could also be the size of the bounding box over time. When a object backs directly away, it may not move much, but it will shrink in size.

**Comment by ozett** [comment]
its a while ago, but i used for a simple case code-snippets from here 

https://www.pyimagesearch.com/2018/07/23/simple-object-tracking-with-opencv/

for AI-detected people tracking with this "centroid-tracking alogrithm", if they walk on the street in front of my cam.

maybe worth considering?

**Reply by blakeblackshear** [reply]
This is how object tracking is implemented in frigate currently.
